components/block/element.js

```diff
@@ -9,62 +10,112 @@ export default class Block extends LitElement {
   static styles = [style];
 
   static properties = {
-    blockId: { type: Number},
+    loading: {type: Boolean, state: true},
+    blockId: {type: String},
   };
 
-  renderMenu(){
-    return html`
-      <nav>
-        <button @click=${this.edit}>Edit</button>
-        <button @click=${this.moveUp}>Move Up</button>
-        <button @click=${this.moveDown}>Move Down</button>
-        <button @click=${this.delete}>Delete</button>
-      </nav>
-    `;
+  constructor() {
+    super();
+
+    this.loading = false;
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+
+    this.model.on('sidebar_save:success', this.refresh.bind(this));
   }
 
-  edit(e){
-    console.log(e.target.innerText)
-    this.refresh();
+  get layout() {
+    return window.parent.Core.g.layout;
   }
 
-  delete(e) {
-    console.log(e.target.innerText)
-    this.remove();
+  get model() {
+    return (this.cached_model ||= this.layout.blocks.findWhere({
+      id: this.blockId,
+    }));
   }
 
-  moveUp(e) {
-    console.log(e.target.innerText)
-    const prevElement = this.previousElementSibling;
-    isBlock(prevElement) && this.parentNode.insertBefore(this, prevElement)
+  get parent() {
```

Comment:
I think we can just simplfy it to:

```js
get parent() {
  const parentId = this.model.attributes.parent_block_id;
  this.cached_parent ||= document.querySelector(`ngl-block[blockId="${parentId}"]`);

  return this.cached_parent;
}
```

Reply:
While this is much more readable than my code, your suggestion however does not distinguish between the cases of "nothing cached yet" and "cached that there is no parent." The suggested function would re-run `document.querySelector` every time the getter is called on the same non-nested block.

My original implementation used `undefined` to mark an empty cache, and `null` to mark that the getter has already been called, and that no parent was found. This suggestion uses the same values*, but it does not differentiate between them to avoid unnecessarily calling `document.querySelector` again.

I definitely agree my code had readability issues, and using `undefined` and `null` for different purposes is not as clear as I would have liked. I'd appreciate some more guidance on this issue and how to resolve it.

-----
\* `document.querySelector` returns `null` for an empty match, and `undefined || null` is `null`, so the behavior is the same: `cached_parent` is `undefined` for an uninitialized cache, `null` for no parent

Reply:
Thanks for explanation. In that case I suggest something like this:

```js
get parent() {
  const parentId = this.model.attributes.parent_block_id;
  this.cached_parent ||= parentId && document.querySelector(`ngl-block[blockId="${parentId}"]`);

  return this.cached_parent;
}
```

Here we weight of between code complication and optimization. But optimization for this is in microseconds or less, and even if we have it without caching concept it would still be good implementation as this part of code is super fast.

So in my last suggestion we just avoid searching for parent if it's `null` and we are ok with cases when `parent_block_id` is present but element is not.

Reply:
Yep, this looks good to me -- I have added your solution in 3924c4dc6acef5ee59e71d06904db0effaf52eed.